File: packages/client/ios/RNOta/Store/BundleStoreException.swift
import Foundation

/// Typed failures for the Bundle Store (same cases as Android `BundleStoreException`).
public enum BundleStoreException: Error, Equatable, LocalizedError {
  case invalidSlotId(String)
  case slotNotFound(String)
  case slotAlreadyExists(String)
  case slotNotCommitted(String)
  case slotInUse(String)
  case diskFull
  case ioFailure(String)

  public var errorDescription: String? {
    switch self {
    case let .invalidSlotId(id): "Invalid slot id: \(id)"
    case let .slotNotFound(id): "Slot not found: \(id)"
    case let .slotAlreadyExists(id): "Slot already exists: \(id)"
    case let .slotNotCommitted(id): "Slot is not committed (missing \(OtaPaths.bundleFileName)): \(id)"
    case let .slotInUse(id): "Slot is referenced by state and cannot be deleted: \(id)"
    case .diskFull: "Disk full while writing OTA store"
    case let .ioFailure(message): message
    }
  }
}

File: packages/client/android/src/main/java/com/rnota/store/BundleStoreException.kt
package com.rnota.store

/**
 * Typed failures for the Bundle Store.
 * Callers (M3/M5/M6) map these to telemetry; the boot-path resolver never uses this API.
 */
sealed class BundleStoreException(
  message: String,
  cause: Throwable? = null,
) : Exception(message, cause) {
  class InvalidSlotId(
    val slotId: String,
  ) : BundleStoreException("Invalid slot id: $slotId")

  class SlotNotFound(
    val slotId: String,
  ) : BundleStoreException("Slot not found: $slotId")

  class SlotAlreadyExists(
    val slotId: String,
  ) : BundleStoreException("Slot already exists: $slotId")

  class SlotNotCommitted(
    val slotId: String,
  ) : BundleStoreException("Slot is not committed (missing ${OtaPaths.BUNDLE_FILE_NAME}): $slotId")

  class SlotInUse(
    val slotId: String,
  ) : BundleStoreException("Slot is referenced by state and cannot be deleted: $slotId")

  class DiskFull(
    cause: Throwable? = null,
  ) : BundleStoreException("Disk full while writing OTA store", cause)

  class IoFailure(
    message: String,
    cause: Throwable? = null,
  ) : BundleStoreException(message, cause)
}
